src/erlcloud_config.erl

```diff
@@ -0,0 +1,632 @@
+%%
+%% https://docs.aws.amazon.com/config/latest/APIReference/API_DescribeComplianceByResource.html
+%%------------------------------------------------------------------------------
+-spec describe_compliance_by_resource() ->
+    result_paged(map()).
+describe_compliance_by_resource() ->
+    describe_compliance_by_resource(#{}).
+
+-spec describe_compliance_by_resource(Opts :: map()) ->
+    result_paged(map()).
+describe_compliance_by_resource(Opts) ->
+    describe_compliance_by_resource(Opts, default_config()).
+
+-spec describe_compliance_by_resource(Opts :: map(),
+    Config :: aws_config()) ->
+    result_paged(map()).
+describe_compliance_by_resource(Opts, Config) when is_map(Opts) orelse Opts =:= #{} ->
+    Request = #{?COMPLIANCE_TYPES => maps:get(?COMPLIANCE_TYPES, Opts, []),
+        ?LIMIT                    => maps:get(?LIMIT, Opts, ?LIMIT_MAX),
+        ?NEXT_TOKEN_LABEL         => maps:get(?NEXT_TOKEN_LABEL, Opts, <<"">>),
+        ?RESOURCE_ID              => maps:get(?RESOURCE_ID, Opts, null),
+        ?RESOURCE_TYPE            => maps:get(?RESOURCE_TYPE, Opts, null)
+    },
+    case request(Config, "DescribeComplianceByResource", Request) of
+        {ok, Res} ->
+            {ok, Res, get_next_token(Res)};
+        {error,  _} = E -> E
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% DescribeConfigRuleEvaluationStatus
+%%
+%% https://docs.aws.amazon.com/config/latest/APIReference/API_DescribeConfigRuleEvaluationStatus.html
+%%------------------------------------------------------------------------------
+-spec describe_config_rule_evaluation_status() ->
+    result_paged(map()).
+describe_config_rule_evaluation_status() ->
+    describe_config_rule_evaluation_status(#{}).
+
+-spec describe_config_rule_evaluation_status(Opts :: map()) ->
+    result_paged(map()).
+describe_config_rule_evaluation_status(Opts) ->
+    describe_config_rule_evaluation_status(Opts, default_config()).
+
+-spec describe_config_rule_evaluation_status(Opts :: map(),
+    Config :: aws_config()) ->
+    result_paged(map()).
+describe_config_rule_evaluation_status(Opts, Config) when is_map(Opts) orelse Opts =:= #{} ->
+    Request = #{?CONFIG_RULE_NAMES => maps:get(?CONFIG_RULE_NAMES, Opts, []),
+        ?LIMIT                    => maps:get(?LIMIT, Opts, ?LIMIT_MAX),
+        ?NEXT_TOKEN_LABEL         => maps:get(?NEXT_TOKEN_LABEL, Opts, <<"">>)
+    },
+    case request(Config, "DescribeConfigRuleEvaluationStatus", Request) of
+        {ok, Res} ->
+            {ok, Res, get_next_token(Res)};
+        {error,  _} = E -> E
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% DescribeConfigRules
+%%
+%% https://docs.aws.amazon.com/config/latest/APIReference/API_DescribeConfigRules.html
+%%------------------------------------------------------------------------------
+-spec describe_config_rules() ->
+    result_paged(map()).
```

Comment:
for every function that returns `result_paged` I think we should add `describe_something_all` function, for those who just want to get it all :)

Examples where we have it already:
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_autoscaling.erl#L68
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_cloudformation.erl#L81
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_ec2.erl#L1872
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_elb.erl#L216
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_iam.erl#L132
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_rds.erl#L119
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_redshift.erl#L99
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_route53.erl#L150
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_sdb.erl#L279
* https://github.com/erlcloud/erlcloud/blob/b50123a62da79/src/erlcloud_sns.erl#L320

Reply:
btw spekaing of opposite point of view, @motobob posted here 
 https://github.com/erlcloud/erlcloud/pull/476#discussion_r168361548 "no more _all funs", so sound like something to agree on & probably add to best practices section

Reply:
Well okay.  I have to admit that I'm rather stunned by this.  

In my opinion implementing "get all" functions without limits or pagination is a software anti-pattern of gargantuan proportions.  So I confidently went off to Google it, confident that I would find about a million software 101 examples of why you should never, never, do this.  And guess what, I was amazed to find that I couldn't really find a single mention!  

This has left me all rather confused, because decades ago implementing a `describe_something_all` function, without limits or pagination, used to be such an obvious anti-pattern that it would instantly eliminate candidates from consideration in a job interview.  Now I'm uncertain as to whether our industry just got amnesia along the way or somewhere along the line I fell asleep.  Now possibly I'm a software dinosaur and I'm waking up to a new reality were all limitations on memory are considered fictional these days and crashing the system because of OOM conditions based on the arguments passed, without defensive programming, is now devolved to be the responsibility of calling client modules as a best practice.  I just hope I'm not the guy on call when the first customer has enough `config` resources to crash the CE process passes the critical threshold! 

Still, I realize that an even worse sin is not following the patterns and spirit of implementation of a software module.  Even if it's written in Swahili, maintainers should stick to Swahili.  So if you really, really, want to, I will hold my nose and implement  `describe_something_all` everywhere.  But you are going to have to justify it to me just one more time.

Reply:
I get it.  Except the issue is not whether it's useful or not.  Sure it's useful.  The problem is that it's dangerous.  Mostly it violates elementary principals of https://en.wikipedia.org/wiki/Defensive_programming.   Traditionally, you should never, never, implement a function that is guaranteed to fail in certain conditions based on the arguments you pass (or lack of them).  I can give you a guaranteed instance where this will crash: just give me a customer with more config results than will fit in working storage.  

It's an anti-pattern from so many perspectives they are hard to even list.  Being able to crash a system on demand is bad from a security perspective too, for a long list of reasons.

This pattern is also vaguely reminiscent of the https://docs.microsoft.com/en-us/azure/architecture/antipatterns/extraneous-fetching/ anti-pattern.

As I said, if everyone insists on this (and given the precedent in the module) I'll bow to the local zeitgeist and implement it.  But it churns my gut.  I'll just add it to a growing list of software war stories I can tell at the bar one day to old has-beens like myself.

Reply:
@kkuzmin @motobob What's your take on this?

Reply:
@ikemsley @alexturkin I totally agree that having _all is an anti-pattern. These functions were introduced as convenient utils that one can use at his own risk accepting possible OOMs.
I like the idea of adding `erlcloud_util:process_all()` which accepts a user defined callback to process a result page. But this approach requires all `describe_*`/`list_*` function to have  common pagination interface.